src/nvmet.c

```diff
@@ -905,36 +881,31 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
 		/* If a host is defined without any port number, just remove and exit. */
 		if (port == INVALID_VALUE)
 			return SUCCESS;
-	} else {
-		/* If no host is defined, remove all hosts */
-		sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-		if ((err = scandir(path, &list, NULL, NULL)) < 0) {
-			goto host_check_out;
-		}
+	}
+
+	/*
+	 * Grab number of entries in allowed_hosts directory. If more than zero still exist, we will not
+	 * proceed with the rest
+	 */
+	sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+	if ((err = scandir(path, &list, NULL, NULL)) < 0) {
+		msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
+		print_error(msg, return_message, path, __func__, strerror(errno));
+		return rc;
+	}
+	hostno = err;
```

Comment:
I believe there is a problem here. `err` will be always >= 2, because `.` and `..` are always present.

At line 896 `hostno` is set the value of `err`, so is always > 0.
Every subsequent checks that require `hostno` to be == 0, will fail. This leads to skip some phases of the unexporting process.

Check at line 908 will always fail. The same happens at line 934.

Here an example of what happens:

```bash
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        None.


Exported NVMe Ports

        None.

shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -g -j -e -b rd2 -P 1000 -H nqn.host1 | jq
{
  "status": "Success",
  "message": "Block device rd2 has been mapped to nqn.host1 through port 1000 as nqn.2021-06.org.rapiddisk:ubuserver-rd2."
}
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2         Namespace: 1    Device: /dev/rd2        Enabled: True

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)     NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -g -j -x -b rd2 -P 1000 -H nqn.host1 | jq
{
  "status": "Success"
}
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -n
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

NVMe Target Exports

        1: NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2         Namespace: 1    Device: /dev/rd2        Enabled: True

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)     NQN: nqn.2021-06.org.rapiddisk:ubuserver-rd2
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$ sudo ./rapiddisk_debug -N
rapiddisk 9.0.0
Copyright 2011 - 2023 Petros Koutoupis

Exported NVMe Ports

        1: Port: 1000 - 192.168.20.40 (tcp)
shub@ubuserver:~/rapiddisk_clion_2_debug_new_branch/src$
```

I tried implementing a `scandir` filter like this, which causes `..` and `.` to be removed from the result passed to `scandir`:

```C
int scandir_filter_no_dot(const struct dirent *list) {
	if ((strncmp(list->d_name, "..", 2) == SUCCESS) || (strncmp(list->d_name, ".", 1) == SUCCESS)) {
		return FALSE;
	}
	return TRUE;
}
```

Passing it as `selector` to the scandir call at line 891, cause `err` to be 0 if nothing except `..` and `.` is found in the path. This approach can lead to avoid the loop at line 897, and gives a chance to checks at lines 908 and 934 to succeed.

Maybe I am missing something, your knowledge is needed here!

Reply:
Wow. I completely overlooked `.` and `..`. I do not know what I was thinking. Nice catch. Looking at your examples, it would be more useful to see it without the `-j`.

I imagine that in your example, because you specified a -H, it removed the mapping from to the hostnqn and did not remove the target or namespace from the filesystem.

Let me look into addressing this and I will probably need to write out a document outlining expectations with command execution.

Reply:
The goal is to make sure that there are no more connected hosts before ultimately removing the port mapping, namespace and target from the subsystem. So, I should have done this from the beginning:
```
diff --git a/src/nvmet.c b/src/nvmet.c
index 533bcae..2792dc5 100644
--- a/src/nvmet.c
+++ b/src/nvmet.c
@@ -891,10 +891,10 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
                print_error(msg, return_message, path, __func__, strerror(errno));
                return rc;
        }
-       hostno = err;
        if (err > 0) {
                for (n = 0; n < err; n++) {
                        if (strncmp(list[n]->d_name, ".", 1) != SUCCESS) {
+                               hostno++;
                                msg = "The target is still mapped to host NQN %s.";
                                print_error(msg, return_message, list[n]->d_name);
                        }
```
That way we avoid counting anything else in the `allowed_hosts` directory.

But I still do need to document this thoroughly to show expectations when certain arguments and conditions are present during command execution.

Reply:
This is another approach:

```diff
diff --git a/src/nvmet.c b/src/nvmet.c
index f91e3ab..a95abf6 100644
--- a/src/nvmet.c
+++ b/src/nvmet.c
@@ -802,6 +802,20 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
        return SUCCESS;
 }

+/**
+ * > If the file name starts with a dot, return false. Otherwise, return true
+ *
+ * @param list This is the directory entry that is being passed to the function.
+ *
+ * @return the number of files in the directory.
+ */
+int scandir_filter_no_dot(const struct dirent *list) {
+       if (strncmp(list->d_name, ".", 1) == SUCCESS) {
+               return FALSE;
+       }
+       return TRUE;
+}
+
 /*
  * description: Unexport an NVMe Target.
  */
@@ -886,19 +900,16 @@ int nvmet_unexport_volume(char *device, char *host, int port, char *return_messa
         * proceed with the rest
         */
        sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
-       if ((err = scandir(path, &list, NULL, NULL)) < 0) {
+       if ((err = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
                msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
                print_error(msg, return_message, path, __func__, strerror(errno));
                return rc;
        }
-       hostno = err;
-       if (err > 0) {
-               for (n = 0; n < err; n++) {
-                       if (strncmp(list[n]->d_name, ".", 1) != SUCCESS) {
-                               msg = "The target is still mapped to host NQN %s.";
-                               print_error(msg, return_message, list[n]->d_name);
-                       }
-               }
+
+       for (n = 0; n < err; n++) {
+               hostno++;
+               msg = "The target is still mapped to host NQN %s.";
+               print_error(msg, return_message, list[n]->d_name);
        }
        list = clean_scandir(list, err);

```
```

Reply:
That is fine: I can add the `scandir_filter_no_dot` but I may add it in another file like the utils file because it can and may be used elsewhere in the future. I will push new code today.